File: spring-boot-web/src/main/java/com/chudichen/spark/web/dao/CourseSearchCountDao.java
package com.chudichen.spark.web.dao;

import com.chudichen.spark.web.domain.CourseClickCount;
import com.chudichen.spark.web.util.HBaseUtil;
import org.apache.commons.compress.utils.Lists;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.ResultScanner;
import org.apache.hadoop.hbase.util.Bytes;
import org.springframework.stereotype.Repository;
import org.springframework.util.CollectionUtils;

import java.util.List;

/**
 * 实战课程访问量Dao层
 *
 * @author chudichen
 * @date 2020-09-07
 */
@Repository
public class CourseSearchCountDao {

    private static final String TABLE = "course_search_click";
    private static final String CF = "info";
    private static final String COLUMN = "click_count";

    public List<CourseClickCount> query(String rowKey) {
        ResultScanner scan = HBaseUtil.INSTANCE.scan(TABLE, rowKey);
        List<CourseClickCount> list = Lists.newArrayList();
        if (scan == null) {
            return list;
        }

        for (Result result : scan) {
            CourseClickCount course = new CourseClickCount();

            String key = Bytes.toString(result.getRow());
            int index = key.lastIndexOf("_");
            String name = key.substring(9,index);
            Long value = Bytes.toLong(result.getValue(Bytes.toBytes(CF), Bytes.toBytes(COLUMN)));

            course.setValue(value);
            course.setName(name);
            list.add(course);
        }
        return list;
    }
}

File: spark-streaming-project/src/main/scala/com/chudichen/spark/dao/CourseSearchCountDao.scala
package com.chudichen.spark.dao

import com.chudichen.spark.domain.CourseSearchCount
import com.chudichen.spark.utils.HBaseUtil

import scala.collection.Iterator
import scala.collection.mutable.ListBuffer

/**
 * 实战课程访问量DAO层
 *
 * @author chudichen
 * @date 2020-09-02
 */
object CourseSearchCountDao {

  val table = "course_search_click"
  val cf = "info"
  val columnQualifier = "click_count"

  /**
   * 向数据库中保存数据
   *
   * @param courseSearchList 课程列表
   */
  def save(courseSearchList: ListBuffer[CourseSearchCount]): Unit = {
    courseSearchList.foreach(course => {
      val rowKey = course.daySearchCourse
      val value = course.clickCount
      HBaseUtil.INSTANCE.putData(table, rowKey, cf, columnQualifier, value)
    })
  }

  def get(daySearchCourse: String): Long = {
    HBaseUtil.INSTANCE.getDataLong(table, daySearchCourse, cf, columnQualifier)
  }

  def increase(list: ListBuffer[CourseSearchCount]): Unit = {
    list.foreach(courseSearchCount => {
      val count = get(courseSearchCount.daySearchCourse) + courseSearchCount.clickCount
      HBaseUtil.INSTANCE.putData(table, courseSearchCount.daySearchCourse, cf, columnQualifier, count)
    })
  }
}
